azurelinuxagent/ga/exthandlers.py

```diff
@@ -469,31 +465,48 @@ def handle_ext_handlers(self, etag=None):
             return
 
         wait_until = datetime.datetime.utcnow() + datetime.timedelta(minutes=_DEFAULT_EXT_TIMEOUT_MINUTES)
-        max_dep_level = max([handler.sort_key() for handler in self.ext_handlers.extHandlers])
 
-        self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
-        for ext_handler in self.ext_handlers.extHandlers:
-            handler_success = self.handle_ext_handler(ext_handler, etag)
+        all_extensions = []
+        for handler in self.ext_handlers.extHandlers:
+            if any(handler.properties.extensions):
+                all_extensions.extend([(ext, handler) for ext in handler.properties.extensions])
+            else:
+                # We need to process the Handler even if no settings specified from CRP (legacy behavior)
+                logger.info("No extension/run-time settings settings found for {0}".format(handler.name))
+                all_extensions.append((None, handler))
+
+        def get_dependency_level(tup):
+            (ext, handler_) = tup
+            if ext is not None:
+                return ext.dependency_level_sort_key(handler_.properties.state)
+            return handler_.dependency_level_sort_key()
+
+        all_extensions.sort(key=get_dependency_level)
+        # Since all_extensions are sorted based on sort_key, the last element would be the maximum based on the sort_key
+        max_dep_level = get_dependency_level(all_extensions[-1]) if any(all_extensions) else 0
+
+        for extension, ext_handler in all_extensions:
+            extension_success = self.handle_ext_handler(ext_handler, extension, etag)
 
             # Wait for the extension installation until it is handled.
             # This is done for the install and enable. Not for the uninstallation.
             # If handled successfully, proceed with the current handler.
             # Otherwise, skip the rest of the extension installation.
-            dep_level = ext_handler.sort_key()
+            dep_level = get_dependency_level((extension, ext_handler))
             if 0 <= dep_level < max_dep_level:
 
                 # Do no wait for extension status if the handler failed
-                if not handler_success:
-                    msg = "Handler: {0} processing failed, will skip processing the rest of the extensions".format(
-                        ext_handler.name)
-                    add_event(AGENT_NAME,
-                              version=CURRENT_VERSION,
-                              op=WALAEventOperation.ExtensionProcessing,
+                if not extension_success:
```

Comment:
The code for depends-on still needs to be reviewed properly, can ignore this code section for now